server/application/websocket_test.go

```diff
@@ -164,3 +165,147 @@ func TestValidateWithoutPermissions(t *testing.T) {
 
 	testServerConnection(t, validate, true)
 }
+
+func TestTerminalSession_Read(t *testing.T) {
+	validate := func(w http.ResponseWriter, r *http.Request) {
+		ts := newTestTerminalSession(w, r)
+		if ts.wsConn == nil {
+			t.Fatalf("WebSocket connection is not initialized")
+		}
+
+		ts.terminalOpts = &TerminalOptions{DisableAuth: true}
+
+		testMessages := []TerminalMessage{
+			{
+				Operation: "stdin",
+				Data:      "test input",
+			},
+			{
+				Operation: "resize",
+				Cols:      80,
+				Rows:      24,
+			},
+			{
+				Operation: "unknown",
+			},
+		}
+
+		for _, msg := range testMessages {
+			bytes, _ := json.Marshal(msg)
+			err := ts.wsConn.WriteMessage(websocket.TextMessage, bytes)
+			if err != nil {
+				t.Errorf("Failed to write message: %v", err)
+			}
+		}
+	}
+
+	tests := []struct {
+		name           string
+		expectedOutput string
+		expectedSize   *remotecommand.TerminalSize
+		expectedError  bool
+	}{
+		{
+			name:           "stdin operation",
+			expectedOutput: "test input",
+			expectedSize:   nil,
+			expectedError:  false,
+		},
+		{
+			name:           "resize operation",
+			expectedOutput: "", // Expecting no output for resize
+			expectedSize: &remotecommand.TerminalSize{
+				Width:  80,
+				Height: 24,
+			},
+			expectedError: false,
+		},
+		{
+			name:           "unknown operation",
+			expectedOutput: EndOfTransmission,
+			expectedSize:   nil,
+			expectedError:  true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```

Comment:
You need to specify a new local value for the testcase inside this loop as t.Run() runs in a different goroutine. Without it, it will just execute the first test you specified in the slice.
A usually used trick is to add a line right after the `for` statement with `tt := tt`. This will generate a new local `tt` variable for each iteration.